astropy/io/fits/src/compressionmodule.c

```diff
@@ -1063,6 +1063,24 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
     if (PyErr_Occurred()) {
         return NULL;
     }
+    // non-integer scaling or subtractive dithering means floating-point
+    if (floor(fileptr->Fptr->cn_bscale) != fileptr->Fptr->cn_bscale ||
```

Comment:
Yes, I think that's somewhat better although I'm still not convinced that this is correct. It could be a float that is bigger than the largest 64bit integer and pass this test.

I'm really wondering: Is there no way to safely check if a `double` is also a valid `long` in C?

Reply:
I remember that the same test in cfitsio is more involved than what I've written here, probably for the reason you've mentioned. I'll expand it.